### PR TITLE
[5.5] Create a new "splitBy" method on the collection

### PR DESCRIPTION
**Pull request by ahmedash95**
Creates a new ``` splitBy ``` method, to split all items in the collection based on a condition:

for example : 

``` php
$emails = collect([

        ["name" => "User 1" , "email" => "user1@laravel.com"],
        ["name" => "User 2" , "email" => "user2@gmail.com"],
        ["name" => "User 3" , "email" => "user3@laracasts.com"],
        ["name" => "User 4" , "email" => "user4@laravel.com"],

])->splitBy(function($row){
    return strpos($row['email'],'@laravel') !== false;
})
```

the result : 

```
array(
   [
        ["name" => "User 1" , "email" => "user1@laravel.com"],
        ["name" => "User 4" , "email" => "user4@laravel.com"],
    ],
    [
        ["name" => "User 2" , "email" => "user2@gmail.com"],
        ["name" => "User 3" , "email" => "user3@laracasts.com"],
    ]
);
```

This method will work like groupBy but based on value, so I might do splitBy and collapse to manipulate the collection of specific data before taking action on them instead of using filter then merge filtered and unfiltered to group them back together

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -1088,6 +1088,22 @@ public function split($numberOfGroups)
     }
 
     /**
+     * Split a collection based on condition into groups.
+     *
+     * @param  closure  $callback
```
Comment (lucasmichot) [inline src/Illuminate/Support/Collection.php]: `@param  \Closure  $callback`

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -1088,6 +1088,22 @@ public function split($numberOfGroups)
...
+     * @return static
+     */
+    public function splitBy($callback)
+    {
+        $results = $this->reduce(function ($results, $row) use ($callback) {
+            $index = $callback($row) ? 0 : 1;
+            $results[$index][] = $row;
+            return $results;
+         }, []);
+        return new static($results);
```
Comment (lucasmichot) [inline src/Illuminate/Support/Collection.php]: Line break before `return`

Comment (lucasmichot) [inline src/Illuminate/Support/Collection.php]: For some reason StyleCI has not run, ping @GrahamCampbell

Comment (ahmedash95) [inline src/Illuminate/Support/Collection.php]: @lucasmichot I also feel strange when I didn't see it running, It's working now.

Thanks

File: tests/Support/SupportCollectionTest.php [hunk 3]
```diff
@@ -1806,6 +1806,20 @@ public function testSplitEmptyCollection()
         );
     }
 
+    public function testSplitBy()
+    {
+        $collection = (new Collection([
+                ['name' => 'User 1', 'email' => 'user1@gmail.com'],
+                ['name' => 'Taylor Otwell', 'email' => 'taylor@laravel.com'],
+            ]))->splitBy(function ($row) {
+                return strpos($row['email'], '@gmail') !== false ? true : false;
```
Comment (lucasmichot) [inline tests/Support/SupportCollectionTest.php]: `return strpos($row['email'], '@gmail') !== false;` is just fine.

File: tests/Support/SupportCollectionTest.php [hunk 4]
```diff
@@ -1806,6 +1806,20 @@ public function testSplitEmptyCollection()
...
+            });
+
+        $this->assertEquals([
```
Comment (lucasmichot) [inline tests/Support/SupportCollectionTest.php]: What about testing `$collection[1]` ?

File: src/Illuminate/Support/Collection.php (hunk 1 above)
Comment (paulofreitas) [inline src/Illuminate/Support/Collection.php]: Maybe `callabe` fits better... :)

File: tests/Support/SupportCollectionTest.php (hunk 4 above)
Comment (ahmedash95) [inline tests/Support/SupportCollectionTest.php]: I did it, Thanks

File: tests/Support/SupportCollectionTest.php (hunk 3 above)
Comment (ahmedash95) [inline tests/Support/SupportCollectionTest.php]: I did it, Thanks

**Comment by taylorotwell** [comment]
This looks exactly like the `partition` method.

**Comment by ahmedash95** [comment]
I didn't notice the ``` partition ``` method in Collection Class , Sorry :).

**Pull request closed without merging**